splink/diagnostics.py

```diff
@@ -0,0 +1,124 @@
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.session import SparkSession
+from pyspark.context import SparkContext, SparkConf
+from pyspark.sql import SQLContext
+
+from typing import Callable
+
+from pyspark.ml.regression import LinearRegression
+from pyspark.ml.linalg import DenseVector
+from pyspark.ml.linalg import Vectors
+from pyspark.ml.evaluation import RegressionEvaluator
+
+from pyspark.sql.types import StructType, StructField, StringType, DoubleType, Row
+import pyspark.sql.functions as f
+from pyspark.sql.functions import when
+from .check_types import check_types
+
+
+import pandas as pd
+import warnings
+
+
+altair_installed = True
+try:
+    import altair as alt
+except ImportError:
+    altair_installed = False
+
+
+@check_types
+def _splink_score_histogram(
+    df_e: DataFrame,
+    spark: SparkSession,
+    splits=[0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95],
+    adjusted=None,
+):
+
+    """splink score histogram diagnostic 
+    
+    
+        Args:
+    
+        
+            df_e (DataFrame): A dataframe of record comparisons containing a splink score, 
+            e.g. as produced by the .... function
+            
+            spark (SparkSession): SparkSession object
+            
+            adjusted : is the score tf adjusted or not? defaults to None
+            
+            splits : list of splits for histogram bins. Has [0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95] for default
+            
+        
+        Returns:
+            
+            (DataFrame) : pandas dataframe of histogram bins for appropriate splink score variable ready to be plotted.
+            
+        
+    """
+
+    # ensure all values are included in histogram
+
+    if splits[0] != 0:
+        splits = [0.0] + splits
+
+    if splits[-1] != 1.0:
+        splits = splits + [1.0]
+
+    # If adjusted is left empty as the default then tf_adjusted_match_prob is used. Otherwise the variable in adjusted is used
+
+    if adjusted == None:
+        hist = (
+            df_e.select("tf_adjusted_match_prob")
+            .rdd.flatMap(lambda x: x)
+            .histogram(splits)
+        )
+    else:
+        hist = df_e.select(adjusted).rdd.flatMap(lambda x: x).histogram(splits)
+
+    hist[1].append(None)
+    hist_df = pd.DataFrame({"splink_score_bin_low": hist[0], "count_rows": hist[1]})
```

Comment:
Need to normalise the count into a frequency density to account for the possibility of variying bin widths.
https://www.me.psu.edu/cimbala/me345/Lectures/Probability_density_functions.pdf

This is equivalent to doing `pyplot.hist(sample, bins=10, density=True)` in matplotlib - see [here](https://machinelearningmastery.com/probability-density-estimation/)
> Importantly, we can convert the counts or frequencies in each bin of the histogram to a normalized probability to ensure the y-axis of the histogram matches the y-axis of the line plot

Once normalised, the total area under the histogram should be equal to 1.

The y axis is then 'Probability density' and the title 'Estimated probability density' or something like that